server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -1892,7 +1925,11 @@ public UserVm upgradeVirtualMachine(ScaleVMCmd cmd) throws ResourceUnavailableEx
         }
         CallContext.current().setEventDetails("Vm Id: " + vm.getUuid());
 
-        boolean result = upgradeVirtualMachine(vmId, newServiceOfferingId, cmd.getDetails());
+        Map<String, String> cmdDetails = cmd.getDetails();
+
+        updateInstanceDetails(cmdDetails, vm, newServiceOfferingId);
+
+        boolean result = upgradeVirtualMachine(vmId, newServiceOfferingId, cmdDetails);
```

Comment:
Sorry for late review.

@GaOrtiga 
if the upgrade fail, should the details be restored to the original values ?

ping @DaanHoogland @GutoVeronezi , anyone has tested it ?

Reply:
> ping @DaanHoogland @GutoVeronezi , anyone has tested it ?

Yes, it was tested.

> if the upgrade fail, should the details be restored to the original values ?

The PR does not address changes to this behavior. it remains the same; we can discuss it in an issue or another PR.

Reply:
> > ping @DaanHoogland @GutoVeronezi , anyone has tested it ?
> 
> Yes, it was tested.
> 
Ok, I asked it because I did not see a comment mentioning it has been manually tested ok.  If not, it would be good to be manually tested.
Why has tested it, except the author?

> > if the upgrade fail, should the details be restored to the original values ?
> 
> The PR does not address changes to this behavior. it remains the same; we can discuss it in an issue or another PR.
> 
> 
Do you think it is a regression of this PR we should address?

Reply:
> Why has tested it, except the author?

I don't think I understood your question.

> Do you think it is a regression of this PR we should address?

The PR does not address changes to method `upgradeVirtualMachine`; therefore, the topic you are questioning would not be a regression of this PR.

Reply:
> > Why has tested it, except the author?
> 
> I don't think I understood your question.
> 
Sorry typo. 
Who has tested it, except the author?

> > Do you think it is a regression of this PR we should address?
> 
> The PR does not address changes to method `upgradeVirtualMachine`; therefore, the topic you are questioning would not be a regression of this PR.
> 

This PR updates the vm instance details, but not restore the vm details if upgrade fails. It should be restored if catch an exception or result is false.
In the past, vm details are not updated so not need to restore in case of failures.

Reply:
> Who has tested it, except the author?

I tested it; I do not know if anybody else has tested it too.

> please ignore my question regarding the vm details, They are not saved in database before upgrade, so no need to restore.

Yes. The only thing this PR does is facilitate calling the API. For instance, without this PR, if one has a VM with 1 vCPU and 1 GB RAM and wants to scale it to 2 GB RAM, one has to call the API with both the current vCPU count and the new RAM size. With this PR, one will only have to pass the new RAM size in the parameters; ACS will retrieve the current VM's vCPU count and will add it to the `details` map to be updated; thus, passing the current vCPU count and the new RAM size will have the same behavior as just passing the new RAM size. It just updates the `details` map (if necessary) before scaling the VM. This applies to RAM, frequency and vCPU.

I created PR #8673 to improve the methods names and docs.